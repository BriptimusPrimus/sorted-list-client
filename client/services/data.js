import config from '../../config';

const apiEndpoint = config.data_service.api.protocol + '://'
  + config.data_service.api.host + ':'
  + config.data_service.api.port + '/'
  + config.data_service.api.root + '/';

function getRequest(endpoint) {
  function makeRequest(resolve, reject) {
    const requestConf = {
      method: 'GET',
      mode: 'cors',
      headers: new Headers({
        'Content-Type': 'application/json'
      })
    }

    fetch(endpoint, requestConf)
      .then(function(response) {
        if (response.status !== 200) {
          return {
            error: 'Service Failure',
            status: response.status
          }
        }
        return response.json();
      })
      .then(function(data) {
        if(data.error) {
          reject(data);
        } else {
          resolve(data);
        }
      })
      .catch(function(reason) {
        reject({
          error: 'Service Failure'
        });
      });
  }

  return new Promise(function(resolve, reject) {
    makeRequest(resolve, reject);
  });
}

function postRequest(enpoint, data) {
  function makeRequest(resolve, reject) {
    const requestConf = {
      method: 'POST',
      mode: 'cors',
      headers: new Headers({
        'Content-Type': 'application/json'
      }),
      body: JSON.stringify(data)
    }

    fetch(endpoint, requestConf)
      .then(function(response) {
        if (response.status !== 200) {
          return {
            error: 'Service Failure',
            status: response.status
          }
        }        
        return response.json();
      })
      .then(function(data) {
        if(data.error) {
          reject(data);
        } else {
          resolve(data);
        }
      })
      .catch(function(reason) {
        reject({
          error: 'Service Failure'
        });
      });
  }

  return new Promise(function(resolve, reject) {
    makeRequest(resolve, reject);
  });

}

function getEmployees() {
  const endpoint = apiEndpoint
    + config.data_service.api.employees_data +
    '?count=' + config.data_service.defaulCount;

  return getRequest(endpoint);
}

export default {
  getEmployees
}