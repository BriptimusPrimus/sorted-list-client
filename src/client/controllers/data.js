import dataService from '../services/data';
import {sortByColumn} from '../actions';

function sortBy(dispatchFn, column) {
  const fetchOpts = {
    sortBy: store.getState().sortBy
  };  
  
  dispatchFn(sortByColumn(column));
  dataService.getEmployees(fetchOpts)
    .then(function fullfilled(data) {
      store.dispatch(receiveData(data));
    })
    .catch(function rejected(reason) {
      console.log('response error:', reason);
    })
}

export default {
  sortBy
}