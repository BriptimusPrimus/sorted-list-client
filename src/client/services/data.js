import fetch from 'isomorphic-fetch';
import querystring from 'querystring';
import config from '../../../config';

const apiEndpoint = `${config.data_service.api.protocol}://${config.data_service.api.host}:${config.data_service.api.port}/${config.data_service.api.root}/`;

const getRequest = async function getRequest(endpoint) {
  const requestConf = {
    method: 'GET',
    mode: 'cors',
    headers: new Headers({
      'Content-Type': 'application/json'
    })
  };

  const response = await fetch(endpoint, requestConf);

  if (response.status === 404) {
    return {
      notFound: true
    };
  }

  if (response.status !== 200) {
    return {
      error: 'Service Failure',
      status: response.status
    };
  }

  const data = await response.json();
  if (data.error) {
    throw new Error(data);
  }

  return data;
};

const getEmployees = function getEmployees(options = {}) {
  const sortBy = options.sortBy || {
    column: 'codeNumber'
  };

  const params = {
    count: options.count || config.data_service.defaulCount,
    sort: sortBy.column,
    order: sortBy.orderDesc ? 'DESC' : 'ASC'
  };

  const endpoint = `${
    apiEndpoint + config.data_service.api.employees_data
  }?${querystring.stringify(params)}`;

  return getRequest(endpoint);
};

const getCustomers = function getCustomers(options = {}) {
  const sortBy = options.sortBy || {
    column: 'id'
  };

  const params = {
    count: options.count || config.data_service.defaulCount,
    sort: sortBy.column,
    order: sortBy.orderDesc ? 'DESC' : 'ASC'
  };

  const endpoint = `${
    apiEndpoint + config.data_service.api.customers_data
  }?${querystring.stringify(params)}`;

  return getRequest(endpoint);
};

const getCustomerDetails = function getCustomerDetails({ customerId } = {}) {
  const endpoint = `${
    apiEndpoint + config.data_service.api.customer_details
  }/${customerId}`;

  return getRequest(endpoint);
};

const getDisplayDetails = function getDisplayDetails({ customerId } = {}) {
  const endpoint = `${
    apiEndpoint + config.data_service.api.display_details
  }?id=${customerId}`;

  return getRequest(endpoint);
};

export { getEmployees, getCustomers, getCustomerDetails, getDisplayDetails };
