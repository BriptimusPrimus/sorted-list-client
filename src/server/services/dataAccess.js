/*
 * Data Access Service Abstraction
 */

const implementationModule = require('../../../config').dal.module;

// eslint-disable-next-line import/no-dynamic-require
const dal = require(`./${implementationModule}`);

const getCustomerDetails = async function getCustomerDetails(id) {
  try {
    const data = await dal.getCustomer(id);

    // Customer not exist
    if (data === undefined) {
      return {
        success: true,
        noDataFound: true
      };
    }

    return {
      success: true,
      data
    };
  } catch (err) {
    return {
      success: false,
      reason: (err && err.message) || err || ''
    };
  }
};

const getCustomers = async function getCustomers(opts) {
  const params = {
    count: Number.isNaN(Number(opts.count)) ? 20 : Number(opts.count),
    page:
      Number.isNaN(Number(opts.page)) || opts.page < 1 ? 1 : Number(opts.page),
    sortBy: opts.sortBy,
    order: opts.order || 'ASC',
    filters: opts.filters || {}
  };

  try {
    const data = await dal.getCustomers(params);
    return {
      success: true,
      data
    };
  } catch (err) {
    return {
      success: false,
      reason: (err && err.message) || err || ''
    };
  }
};

const getEmployees = async function getEmployees(opts) {
  const params = {
    count: Number.isNaN(Number(opts.count)) ? 20 : Number(opts.count),
    page:
      Number.isNaN(Number(opts.page)) || opts.page < 1 ? 1 : Number(opts.page),
    sortBy: opts.sortBy,
    order: opts.order || 'ASC',
    filters: opts.filters || {}
  };

  try {
    const data = await dal.getEmployees(params);
    return {
      success: true,
      data
    };
  } catch (err) {
    return {
      success: false,
      reason: (err && err.message) || err || ''
    };
  }
};

// This object implements the bridge interface:
// interface: {
//   getCustomerDetails: async function(customerId){}
//   getCustomers: async function(options){}
//   getEmployees: async function(options){}
// }
module.exports = {
  getCustomerDetails,
  getCustomers,
  getEmployees
};
