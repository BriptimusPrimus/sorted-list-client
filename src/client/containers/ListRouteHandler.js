import React from 'react';
import PropTypes from 'prop-types';
import { receiveData, receiveDataError } from '../actions';
import { getEmployees } from '../services/data';
import ListContainer from './ListContainer';
import withUniversalLoad from '../lib/withUniversalLoad';

const ListWithUniversalLoad = withUniversalLoad(ListContainer, {
  dataSource: getEmployees,
  dataSourceParams: {
    sortBy: {
      inState: true
    }
  },
  successAction: receiveData,
  failureAction: receiveDataError
});

const ListRouteHandler = ({ initialState }) => {
  return <ListWithUniversalLoad initialState={initialState} />;
};

// Override loadData
ListRouteHandler.loadData = async function preloadData(params) {
  const data = await ListWithUniversalLoad.loadData(params);
  return {
    data
  };
};

export default ListRouteHandler;

ListRouteHandler.propTypes = {
  initialState: PropTypes.shape({
    data: PropTypes.shape({
      list: PropTypes.arrayOf(
        PropTypes.shape({
          firstName: PropTypes.string,
          surname: PropTypes.string,
          surname2: PropTypes.string,
          codeNumber: PropTypes.string,
          rfc: PropTypes.string,
          status: PropTypes.string
        })
      )
    }),
    sortBy: PropTypes.shape({
      column: PropTypes.string,
      orderDesc: PropTypes.bool
    })
  })
};

ListRouteHandler.defaultProps = {
  initialState: {
    data: {
      list: []
    },
    sortBy: {
      column: 'codeNumber',
      orderDesc: false
    }
  }
};
