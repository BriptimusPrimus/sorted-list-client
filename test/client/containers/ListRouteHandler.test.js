import React from 'react';
import { render, cleanup, fireEvent, waitFor } from '@testing-library/react';
import ListRouteHandler from '../../../src/client/containers/ListRouteHandler';
import * as mockService from '../../../src/client/services/data';
import {
  getInitialState,
  sortedByName,
  sortedBySurname,
  sortedByCodeNumberDesc,
  sortedByRFC,
  sortedByStatus
} from '../utils/dataMocks';

jest.mock('react-router-dom', () => {
  return {
    useParams: () => ({}),
    useLocation: () => ({})
  };
});
jest.mock('../../../src/client/services/data');

describe('ListRouteHandler tests', () => {
  const initialState = getInitialState();

  beforeEach(() => {
    jest.clearAllMocks();
    mockService.getEmployees.mockImplementation(async () => {
      return initialState.data;
    });
  });

  afterEach(cleanup);

  test('Renders', () => {
    const { container } = render(
      <ListRouteHandler initialState={initialState} />
    );
    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Sorts column', async () => {
    mockService.getEmployees.mockImplementation(async () => {
      return sortedByName;
    });

    const { getByText, queryByText } = render(
      <ListRouteHandler initialState={initialState} />
    );
    expect(queryByText('ELOINA')).not.toBeNull();

    const nameHeader = getByText('Name');
    fireEvent.click(nameHeader);

    // Wait for AJAX call to be complete
    await waitFor(() => {});

    expect(queryByText('ELOINA')).toBeNull();
    expect(queryByText('ABDI')).not.toBeNull();
  });

  test('Sort column service call fails', async () => {
    mockService.getEmployees.mockImplementation(async () => {
      return Promise.reject(new Error('boom'));
    });

    const { getByText, queryByText } = render(
      <ListRouteHandler initialState={initialState} />
    );
    expect(queryByText('ELOINA')).not.toBeNull();

    const nameHeader = getByText('Name');
    fireEvent.click(nameHeader);

    // Wait for AJAX call to be complete
    await waitFor(() => {});

    expect(queryByText('ELOINA')).not.toBeNull();
    expect(queryByText('ABDI')).toBeNull();
  });

  test('Sorts by surname', async () => {
    mockService.getEmployees.mockImplementation(async () => {
      return sortedBySurname;
    });

    const { getByText, queryByText } = render(
      <ListRouteHandler initialState={initialState} />
    );
    expect(queryByText('SALINAS VAZQUEZ')).not.toBeNull();

    const nameHeader = getByText('Surname');
    fireEvent.click(nameHeader);

    // Wait for AJAX call to be complete
    await waitFor(() => {});

    expect(queryByText('SALINAS VAZQUEZ')).toBeNull();
    expect(queryByText('ABAD MARTINEZ')).not.toBeNull();
  });

  test('Sorts by codeNumber', async () => {
    mockService.getEmployees.mockImplementation(async () => {
      return sortedByCodeNumberDesc;
    });

    const { getByText, queryByText } = render(
      <ListRouteHandler initialState={initialState} />
    );
    expect(queryByText('10001')).not.toBeNull();

    const nameHeader = getByText('Code Number');
    fireEvent.click(nameHeader);

    // Wait for AJAX call to be complete
    await waitFor(() => {});

    expect(queryByText('10001')).toBeNull();
    expect(queryByText('42850')).not.toBeNull();
  });

  test('Sorts by RFC', async () => {
    mockService.getEmployees.mockImplementation(async () => {
      return sortedByRFC;
    });

    const { getByText, queryByText } = render(
      <ListRouteHandler initialState={initialState} />
    );
    expect(queryByText('LUIS RAMON')).not.toBeNull();

    const nameHeader = getByText('RFC');
    fireEvent.click(nameHeader);

    // Wait for AJAX call to be complete
    await waitFor(() => {});

    expect(queryByText('LUIS RAMON')).toBeNull();
    expect(queryByText('ALEJANDRA')).not.toBeNull();
  });

  test('Sorts by Status', async () => {
    mockService.getEmployees.mockImplementation(async () => {
      return sortedByStatus;
    });

    const { getByText, queryByText } = render(
      <ListRouteHandler initialState={initialState} />
    );
    expect(queryByText('Candidate')).not.toBeNull();

    const nameHeader = getByText('Status');
    fireEvent.click(nameHeader);

    // Wait for AJAX call to be complete
    await waitFor(() => {});

    expect(queryByText('Candidate')).toBeNull();
    expect(queryByText('On leave')).not.toBeNull();
  });

  test('Static load data', async () => {
    const result = await ListRouteHandler.loadData();
    expect(result).toEqual({
      data: initialState.data
    });
  });

  test('Renders empty data by default', () => {
    const { container, queryByText } = render(<ListRouteHandler />);
    expect(container.firstChild).not.toBeNull();
    expect(queryByText('ELOINA')).toBeNull();
  });
});
