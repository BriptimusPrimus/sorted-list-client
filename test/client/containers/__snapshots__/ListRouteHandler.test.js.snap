// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListRouteHandler tests Renders 1`] = `
<div
  id="main-view"
>
  <div
    class="panel panel-default row"
  >
    <div
      class="col-xs-0 col-sm-1 col-md-2"
    />
    <section
      class="col-xs-12 col-sm-10 col-md-8"
    >
      <h2
        class="text-primary text-center"
      >
        Sorted List with Data
      </h2>
      <div
        class="well"
      >
        <table
          class="table table-striped table-condensed table-bordered"
        >
          <thead>
            <tr>
              <th
                class="col-xs-3 text-center actionable"
              >
                Name
              </th>
              <th
                class="col-xs-3 text-center actionable"
              >
                Surname
              </th>
              <th
                class="col-xs-2 text-center actionable order-asc"
              >
                Code Number
              </th>
              <th
                class="col-xs-2 text-center actionable"
              >
                RFC
              </th>
              <th
                class="col-xs-2 text-center actionable"
              >
                Status
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                class="text-center"
              >
                ELOINA
              </td>
              <td
                class="text-center"
              >
                SALINAS
                 
                VAZQUEZ
              </td>
              <td
                class="text-center"
              >
                10001
              </td>
              <td
                class="text-center"
              />
              <td
                class="text-center"
              >
                Hired
              </td>
            </tr>
            <tr>
              <td
                class="text-center"
              >
                TERESA DE JESUS
              </td>
              <td
                class="text-center"
              >
                JOFRE
                 
                GARFIAS
              </td>
              <td
                class="text-center"
              >
                10027
              </td>
              <td
                class="text-center"
              >
                JOGT000101
              </td>
              <td
                class="text-center"
              />
            </tr>
            <tr>
              <td
                class="text-center"
              >
                JOSE LUIS
              </td>
              <td
                class="text-center"
              >
                JAVIER
                 
                PECH
              </td>
              <td
                class="text-center"
              >
                10030
              </td>
              <td
                class="text-center"
              >
                JAPL740105
              </td>
              <td
                class="text-center"
              >
                Candidate
              </td>
            </tr>
            <tr>
              <td
                class="text-center"
              >
                ERNESTO
              </td>
              <td
                class="text-center"
              >
                HERNANDEZ
                 
                GARCIA
              </td>
              <td
                class="text-center"
              >
                10031
              </td>
              <td
                class="text-center"
              >
                HEGE521107
              </td>
              <td
                class="text-center"
              >
                Hired
              </td>
            </tr>
            <tr>
              <td
                class="text-center"
              >
                SERGIO
              </td>
              <td
                class="text-center"
              >
                FONSECA
                 
                VIDAL
              </td>
              <td
                class="text-center"
              >
                10137
              </td>
              <td
                class="text-center"
              >
                FOVS820331
              </td>
              <td
                class="text-center"
              >
                Hired
              </td>
            </tr>
            <tr>
              <td
                class="text-center"
              >
                PEDRO ANTONIO
              </td>
              <td
                class="text-center"
              >
                JIMENEZ
                 
                JIMENEZ
              </td>
              <td
                class="text-center"
              >
                10155
              </td>
              <td
                class="text-center"
              >
                JIJP770117
              </td>
              <td
                class="text-center"
              >
                Hired
              </td>
            </tr>
            <tr>
              <td
                class="text-center"
              >
                EFREN
              </td>
              <td
                class="text-center"
              >
                GONZALEZ
                 
                CHAVEZ
              </td>
              <td
                class="text-center"
              >
                10223
              </td>
              <td
                class="text-center"
              >
                GOCE690929
              </td>
              <td
                class="text-center"
              >
                Hired
              </td>
            </tr>
            <tr>
              <td
                class="text-center"
              >
                FRANCISCO
              </td>
              <td
                class="text-center"
              >
                JIMENEZ
                 
                RODRIGUEZ
              </td>
              <td
                class="text-center"
              >
                10337
              </td>
              <td
                class="text-center"
              >
                JIRF651124
              </td>
              <td
                class="text-center"
              >
                Hired
              </td>
            </tr>
            <tr>
              <td
                class="text-center"
              >
                MANUEL
              </td>
              <td
                class="text-center"
              >
                VALDERRABANO
                 
                TACU
              </td>
              <td
                class="text-center"
              >
                10369
              </td>
              <td
                class="text-center"
              >
                VATM701227
              </td>
              <td
                class="text-center"
              >
                Hired
              </td>
            </tr>
            <tr>
              <td
                class="text-center"
              >
                LUIS RAMON
              </td>
              <td
                class="text-center"
              >
                PERALTA
                 
                MAY
              </td>
              <td
                class="text-center"
              >
                10370
              </td>
              <td
                class="text-center"
              >
                PEML740831
              </td>
              <td
                class="text-center"
              >
                Hired
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </section>
    <div
      class="col-xs-0 col-sm-1 col-md-2"
    />
  </div>
</div>
`;
